numpy/_core/src/multiarray/unique.cpp

```diff
@@ -51,7 +52,16 @@ size_t hash_complex(const T *value, npy_bool equal_nan) {
 
     // Now, equal_nan is false or neither of the values is not NaN.
     // So we don't need to worry about NaN here.
-    const unsigned char* value_bytes = reinterpret_cast<const unsigned char*>(value);
+
+    std::complex<S> tmp = *reinterpret_cast<const std::complex<S> *>(value);
+    // Get rid of -0.0 in tmp.
+    if (value_real == 0) {
+        tmp.real(0.0);
+    }
+    if (value_imag == 0) {
+        tmp.imag(0.0);
+    }
+    const unsigned char* value_bytes = reinterpret_cast<const unsigned char*>(&tmp);
     size_t hash = npy_fnv1a(value_bytes, sizeof(T));
```

Comment:
```suggestion
    value_real = (value_real == 0.0) ? NPY_PZERO : value_real;
    value_imag = (value_imag == 0.0) ? NPY_PZERO : value_imag;
    S value_array[2] = {value_real, value_imag};
    const unsigned char* value_bytes = reinterpret_cast<const unsigned char*>(value_array);
    size_t hash = npy_fnv1a(value_bytes, sizeof(S) * 2);
```

Instead of creating a new complex value, it might be more efficient to construct the array (and buffer) directly from the real and imaginary parts after normalizing -0.0 to +0.0.
After applying this change, the benchmark results showed "BENCHMARKS NOT SIGNIFICANTLY CHANGED".

Reply:
Thanks @math-hiyoko, I'll take a look. I was just iterating a bit after my initial quick hacks and noticed some potential optimizations.

Reply:
I tried this, but I got pretty much the same performance penalty for the double precision case (m2 macbook, clang 18.1.8).  Are you sure you committed your suggested change before running the benchmark?

Reply:
> I tried this, but I got pretty much the same performance penalty for the complex double precision case (m2 macbook, clang 18.1.8). Are you sure you committed your suggested change before running the benchmark?

Please try applying the suggested change _and_ removing `#include <complex>` before running the benchmark.
In my environment, the benchmark results changed noticeably depending on whether `#include <complex>` was present.